bb8/src/internals.rs

```diff
@@ -81,6 +109,7 @@ where
     conns: VecDeque<IdleConn<M::Connection>>,
     num_conns: u32,
     pending_conns: u32,
+    inflight_gets: u32,
```

Comment:
Hmm I don't think it's possible, unless there's other state I'm not aware of.

In the following table I'm modeling the internal state following 3 gets, a drop + get, and a notify from the drop. During this time, none of the pending connect tasks have returned with a connection.

| Time | Action | Conns | Idle | Pending | In-Flight |
|--------|--------|--------|--------|--------|--------|
| 0 | - | 10 | 0 | 0 | 0 |
| 1 | Get | 10 | 0 | 1 | 1 |
| 2 | Get | 10 | 0 | 2 | 2 |
| 3 | Get | 10 | 0 | 3 | 3 |
| 4 | Drop | 10 | 1 | 3 | 3 |
| 5 | Get | 10 | 0 | 3 | 3 |
| 6 | Get_Notify | 10 | 0 | 3 | 3 |

In this case, a 4th `Get` is able to cut in front of the in-flight `Get`s, immediately taking the idle connection. When the `Get` at `t1` is notified, it sees nothing in the pool and 3 pending connect tasks. However in isolation it doesn't know that one of those 3 pending are for it, or that it needs to spawn its own connect task.

`num_conns` isn't particularly helpful here since `Get` only cares about what's available in the idle pool. And `pending_conns` by itself isn't enough, since we don't want 3 in-flight `Get`s to be waiting on a single pending connection.

Reply:
Okay, makes sense.